tracer/src/Datadog.Trace/TraceProcessors/TraceUtil.cs

```diff
@@ -0,0 +1,344 @@
+// <copyright file="TraceUtil.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Text;
+using Datadog.Trace.Util;
+
+namespace Datadog.Trace.TraceProcessors
+{
+    internal class TraceUtil
+    {
+        // MaxTagLength the maximum length a tag can have
+        private const int MaxTagLength = 200;
+
+        private static readonly UTF8Encoding Encoding = new UTF8Encoding(false);
+
+        // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/traceutil/truncate.go#L36-L51
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool TruncateUTF8(ref string value, int limit)
```

Comment:
How bad is it to call `Encoding.GetByteCount()` many times on every tag, especially for longer strings like SQL queries? How strict is the tag length limit? What happens is we go over the limit, does it get truncated in the intake anyway or does it get rejected?

Some quick thoughts if we need to optimize this `TruncateUTF8()` 🧠⚡
- The longest Unicode code point is 4 bytes in UTF-8. Can we short-circuit this method if `value.Length < (limit/4)` before we even call `Encoding.GetByteCount()` once? I think it's low-effort/high-reward and we can do it in this PR if you agree.
- UTF-8 is designed to be "self-synchronizing", so if you split a multi-byte character in the middle, it's easy to find where last character started by backing up at most 3 bytes to find the leading byte. It may look something like:
  - truncate string to 200 character (optional)
  - convert to UTF-8 bytes (which could be more than 200 bytes)
  - truncate to 200 bytes
  - to avoid an invalid character at the end, find where the last character started in case we need to drop the last 1-3 bytes
- We could try to make a single iteration pass of the string to count bytes until we reach the limit, instead of calculating the total length of progressively longer strings.
  - `length += Encoding.GetByteCount(valueCharArray, i, 1)`
  - for a string that's 100 characters, we call `Encoding.GetByteCount()` 100 times for 1 character each, instead of calling it 100 times with progressively longer strings (`substring(1)`, `substring(2)`, `substring(3)`, etc)
- Can we fuzz this? If the max length is a little flexible (for example, if we can send 210 bytes without issues because the backend will truncate it again), I think we can optimize for the common scenario (1 byte/char) and always truncate at 200 chars (instead of 200 bytes) and we could be correct most of the time. Or truncate at ~180 and be correct even more often (assuming only a few strings have only a few multi-byte chars). If you think about it, after the first 180 characters in a tag, will anyone miss the next 20 chars?

Reply:
(I think I write too much later in the day)

Reply:
https://gist.github.com/lucaspimentel/b871ee0c911ba24018a2df27143daec8

Reply:
Thank you for this analysis!, TBH I was focusing in a simple port in this PR without looking for these nice optimizations....

I'll apply your `gist` approach for now, and then we can look any other optimization.

Reply:
>How strict is the tag length limit? What happens is we go over the limit, does it get truncated in the intake anyway or does it get rejected?

Only partially related, but if we use this logic to calculate metrics in the tracer (as we need to), then I think we need to make sure we use exactly the same truncation, otherwise we could end up with strange discrepancies? Not entirely sure or thought through yet though 😅

Reply:
done in [7b643b4](https://github.com/DataDog/dd-trace-dotnet/pull/2402/commits/7b643b4bdcc33deb0febaf353471ac302e5b4c63)

Reply:
> we need to make sure we use exactly the same truncation

Good point! If truncation can affect tag-based metrics, truncating algorithm should probably match as well.